relay-server/src/actors/processor.rs

```diff
@@ -936,68 +936,295 @@ impl EnvelopeProcessorService {
-            // the envelope.
+        let Some(ty) = state
+            .event_type()
+            // get the item type from the event type if there is an event
+            .map(ItemType::from_event_type)
+            // othrwise get the first item from the items list and assume it's the only one defined
+            // which we are waiting for.
+            .or_else(|| {
+                state
+                    .envelope()
+                    .items()
+                    .next()
+                    .map(|item| item.ty().clone())
+            })
+        else {
+            relay_log::error!("There are not type and no items in the envelope");
+            return Err(ProcessingError::NoEventPayload);
+        };
 
-            if_processing!({
-                unreal::expand(state, &self.inner.config)?;
-            });
+        // Defined processing pipeilnes based on the event type and/or item type they contain.
+        match ty {
+            // This can still contain attachements.
+            ItemType::Event => {
+                event::extract(state, &self.inner.config)?;
 
-            event::extract(state, &self.inner.config)?;
-            profile::transfer_id(state);
+                if_processing!({
+                    attachment::create_placeholders(state);
+                });
 
-            if_processing!({
-                unreal::process(state)?;
-                attachment::create_placeholders(state);
-            });
+                event::finalize(state, &self.inner.config)?;
+                self.light_normalize_event(state)?;
+                dynamic_sampling::normalize(state);
+                event::filter(state)?;
+                dynamic_sampling::run(state, &self.inner.config);
+                dynamic_sampling::sample_envelope(state)?;
 
-            event::finalize(state, &self.inner.config)?;
-            self.light_normalize_event(state)?;
-            dynamic_sampling::normalize(state);
-            event::filter(state)?;
-            dynamic_sampling::run(state, &self.inner.config);
+                if_processing!({
+                    event::store(state, &self.inner.config, self.inner.geoip_lookup.as_ref())?;
+                    self.enforce_quotas(state)?;
+                });
+
+                if state.has_event() {
+                    event::scrub(state)?;
+                    event::serialize(state)?;
+                }
 
-            // We avoid extracting metrics if we are not sampling the event while in non-processing
-            // relays, in order to synchronize rate limits on indexed and processed transactions.
-            if self.inner.config.processing_enabled() || state.sampling_result.should_drop() {
-                self.extract_metrics(state)?;
+                attachment::scrub(state);
             }
+            // Contains data which belongs together with transactions, e.g. profiles attachments.
+            ItemType::Transaction | ItemType::Profile => {
```

Comment:
Profiles are currently only allowed if they're in the same envelope as a transaction, in which case `ty` is always `Transaction`, right?
```suggestion
            ItemType::Transaction => {
```

Reply:
Yeah, the profile is in there, since I want to cover all the branches in match for `ItemType`